File: leetcode_learning/code_cpp/20.cpp
#include<iostream>
#include<string>
#include<stack>
#include<map>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        //哈希表
        map<char, char> m;
        m[')'] = '(';
        m[']'] = '[';
        m['}'] = '{';
        
        // map<char, char> m = {
        //     {')', '('},
        //     {']', '['},
        //     {'}', '{'}};

        int size = s.size();
        // cout<<size<<endl;
        if(size%2==1)
            return false;
        stack<char> stk;
        for(int i=0;i<size;i++)
        {
            if(stk.empty())
                stk.push(s[i]);
            else
            {
                if(m.count(s[i]))
                {
                    if(stk.top()==m[s[i]])
                        stk.pop();
                    else
                        return false;//如果s[i]是右括号且当前栈顶
                                     //元素不是相应的左括号，则返回false
                }
                else
                    stk.push(s[i]);
            }
        }

        return stk.empty();
    }
};

File: leetcode_learning/code_py/20.py

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s)%2==1:
            return False
        
        match_dict = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
        stack = [] # python的列表为堆栈
        for per_char in s:
            if per_char in match_dict:
                # 需要判断该左括号是否可以闭合
                if len(stack)==0 or stack[-1]!=match_dict[per_char]:
                    # 如果此时堆栈为空，或者堆栈顶不能闭合该左括号
                    return False
                else:
                    # 如果可以闭合，则把栈顶删除
                    stack.pop()
            else:
                # 遇到新的右括号要添加进堆栈
                stack.append(per_char)
        
        # 如果最后栈为空，则返回True
        if len(stack)==0:
            return True
        else:
            return False


if __name__=='__main__':
    s = Solution()
    test_s = '()[]{}'
    print(s.isValid(test_s))
